flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/ResettableExternalBuffer.java

```diff
@@ -0,0 +1,699 @@
+				}
+			}
+			return freeMemory;
+		}
+
+		// Find the index of the first element which is strictly greater than `goal` in `list`.
+		// `list` must be sorted.
+		// If every element in `list` is not larger than `goal`, return `list.size()`.
+		private int upperBound(int goal, List<Integer> list) {
+			if (list.size() == 0) {
+				return 0;
+			}
+			if (list.get(list.size() - 1) <= goal) {
+				return list.size();
+			}
+
+			// Binary search
+			int head = 0;
+			int tail = list.size() - 1;
+			int mid;
+			while (head < tail) {
+				mid = (head + tail) / 2;
+				if (list.get(mid) <= goal) {
+					head = mid + 1;
+				} else {
+					tail = mid;
+				}
+			}
+			return head;
+		}
+	}
+
+	@VisibleForTesting
+	List<ChannelWithMeta> getSpillChannels() {
+		return spilledChannelIDs;
+	}
+
+	/**
+	 * In memory buffer that stores records to memorySegments, returns a iterator that map from memory.
+	 */
+	private class InMemoryBuffer implements Closeable {
+
+		private final int segmentSize;
+		private final ArrayList<MemorySegment> freeMemory;
+		private final AbstractRowSerializer serializer;
+		private final ArrayList<MemorySegment> recordBufferSegments;
+		private final SimpleCollectingOutputView recordCollector;
+
+		// Can't use recordCollector.getCurrentOffset(), maybe the offset of recordCollector is
+		// disrupted by the attempt of record writing.
+		private long currentDataBufferOffset;
+		private int numBytesInLastBuffer;
+
+		private int recordCount;
+
+		private LinkedList<BufferIterator> iterators;
+
+		private InMemoryBuffer(AbstractRowSerializer serializer) {
+			this.segmentSize = memory.get(0).size();
+			this.freeMemory = new ArrayList<>(memory);
+			// serializer has states, so we must duplicate
+			this.serializer = (AbstractRowSerializer) serializer.duplicate();
+			this.recordBufferSegments = new ArrayList<>(memory.size());
+			this.recordCollector = new SimpleCollectingOutputView(this.recordBufferSegments,
+					new ListMemorySegmentSource(this.freeMemory), this.segmentSize);
+			this.recordCount = 0;
+			this.iterators = new LinkedList<>();
+		}
+
+		private void reset() {
+			this.currentDataBufferOffset = 0;
+			this.recordCount = 0;
+
+			// reset free and record segments.
+			this.freeMemory.addAll(this.recordBufferSegments);
+			this.recordBufferSegments.clear();
+
+			this.recordCollector.reset();
+
+			iterators.clear();
+		}
+
+		@Override
+		public void close() {
+			this.freeMemory.clear();
+			this.recordBufferSegments.clear();
+			iterators.clear();
+		}
+
+		public boolean write(BaseRow row) throws IOException {
+			try {
+				this.serializer.serializeToPages(row, this.recordCollector);
+				currentDataBufferOffset = this.recordCollector.getCurrentOffset();
+				numBytesInLastBuffer = this.recordCollector.getCurrentPositionInSegment();
+				recordCount++;
+
+				for (BufferIterator iterator : iterators) {
+					iterator.recordBuffer.updateLimitInLastSegment(numBytesInLastBuffer);
```

Comment:
Why we need to update limit in *last* segment for each write?

Reply:
This is because ResettableExternalBuffer supports reading while writing, and when a Record is written, we need to set the reading iterator to new offset.

Reply:
but not each write will touch the last segment, right? so i think the reader will be confused by it

Reply:
Every write will touch the last segment.
The previous data is on disk, and the last data is in memory.
Each write is append written at the end of InMemoryBuffer.

Reply:
I think we can not support reading while writing now. The current solution does not handle Spill properly and `MergeJoinOperator` in Blink depends on `TwoInputOperatorChain`. I will remove it.
`RecordCount` should be used to solve this problem in the future instead of `LimitInLastSegment`.